File: Session11/207. Course Schedule-indegree.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Graph g=new Graph();
        int indegree[]=new int[numCourses];
        Queue<Integer> q=new LinkedList<>();
        boolean visited[]=new boolean[numCourses];
        for(int p[]:prerequisites)
        {
            ArrayList<Integer> temp=g.adj.getOrDefault(p[1],new ArrayList<>());
            temp.add(p[0]);
            g.adj.put(p[1],temp);
            indegree[p[0]]++;
        }
        
        for(int i=0;i<numCourses;i++)
        {
            if(g.adj.get(i)==null)
            {
                g.adj.put(i,new ArrayList<Integer>());
            }
        }
            
        for(int i=0;i<numCourses;i++)
        {
            if(indegree[i]==0)
            {
                q.add(i);
                visited[i]=true;
            }
        }
        
        while(!q.isEmpty())
        {
            int curr=q.poll();
            //for(int nei:g.adj.getOrDefault(curr,new ArrayList<>()))
            for(int nei:g.adj.get(curr))
            {
                indegree[nei]-=1;
                if(!visited[nei] && indegree[nei]==0)
                {
                    q.add(nei);
                    visited[nei]=true;
                }
            }
        }
        
        for(int i=0;i<numCourses;i++)
        {
            if(visited[i]==false)
            {
                return false;
            }
        }
        return true;
    }

}

class Graph
{
    HashMap<Integer,ArrayList<Integer>> adj;
    Graph()
    {
        adj=new HashMap<>();
    }
}

File: Session11/207. Course Schedule-indegree.py
from collections import defaultdict, deque
class Solution:

    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """

        # Prepare the graph
        adj_list = defaultdict(list)
        indegree = {}
        for dest, src in prerequisites:
            adj_list[src].append(dest)

            # Record each node's in-degree
            indegree[dest] = indegree.get(dest, 0) + 1

        # Queue for maintainig list of nodes that have 0 in-degree
        zero_indegree_queue = deque([k for k in range(numCourses) if k not in indegree])

        topological_sorted_order = []
        count=0
        # Until there are nodes in the Q
        while zero_indegree_queue:

            # Pop one node with 0 in-degree
            vertex = zero_indegree_queue.popleft()
            topological_sorted_order.append(vertex)
            count+=1
            # Reduce in-degree for all the neighbors
            if vertex in adj_list:
                for neighbor in adj_list[vertex]:
                    indegree[neighbor] -= 1

                    # Add neighbor to Q if in-degree becomes 0
                    if indegree[neighbor] == 0:
                        zero_indegree_queue.append(neighbor)

        return True if count == numCourses else False